### PR TITLE
Create Gibbs estimator + tests

### PR DESCRIPTION
**Pull request by laurgao**
- Implement the Gibbs objective function from https://arxiv.org/pdf/1909.07621.pdf
- Move estimators to their own folder

### CODE REVIEW

File: src/python/zquantum/qaoa/estimators/gibbs_estimator.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+from zquantum.core.interfaces.backend import QuantumBackend
+from zquantum.core.measurement import ExpectationValues, Measurements
+from zquantum.core.bitstring_distribution import BitstringDistribution
+from zquantum.core.interfaces.estimation import (
+    EstimateExpectationValues,
+    EstimationTask,
+)
+
+
+class GibbsEstimator(EstimateExpectationValues):
```
Comment (mstechly) [inline src/python/zquantum/qaoa/estimators/gibbs_estimator.py:14]: I'd rename it to `GibbsObjectiveEstimator`.

File: tests/zquantum/qaoa/estimators/gibbs_estimator_test.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    @pytest.fixture()
+    def estimation_tasks(self, operator, circuit):
+        return [EstimationTask(operator, circuit, 10)]
+
+    @pytest.fixture()
+    def backend(self, request):
+        backend = MockQuantumBackend()
+
+        def custom_run_circuit_and_measure(circuit, n_samples):
```
Comment (mstechly) [inline tests/zquantum/qaoa/estimators/gibbs_estimator_test.py:36]: We've recently updated how `MockQuantumBackend` works, this whole function is no longer needed, you can just do: `return MockQuantumBackend()`.

File: tests/zquantum/qaoa/estimators/gibbs_estimator_test.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    ):
+        # Given
+        estimation_tasks = [EstimationTask(QubitOperator("X0"), circuit, 10)]
+        with pytest.raises(TypeError):
+            estimator(
+                backend=backend,
+                estimation_tasks=estimation_tasks,
+            )
+
+    def test_gibbs_estimator_raises_exception_if_alpha_less_than_0(
```
Comment (mstechly) [inline tests/zquantum/qaoa/estimators/gibbs_estimator_test.py:54]: It's good to add test cases for boundary conditions as well – in this case for `alpha=0`.

**Review by mstechly — CHANGES_REQUESTED** [review]

**Comment by alexjuda** [comment]
@laurgao @mstechly what do you folks think about renaming `zquantum.qaoa.estimators.cvar_estimator` into `zquantum.qaoa.estimators._cvar_estimator` and thus separating public/private API by a convention? Anything in `zquantum.qaoa.estimators` would be treated as the public (the functions/classes that we indend our users to freely use in their code). 

BTW this is a convention we started to introduce some time ago in z-quantum-core.

**Comment by mstechly** [comment]
@alexjuda good idea!

**Comment by alexjuda** [comment]
@laurgao If you also like this idea it'd be also nice to follow this convention for public/private API split in #56 🙂

**Comment by laurgao** [comment]
@alexjuda Cool, I'll make the changes. One concern though: for #56, I've seen Farhi ansatz commonly imported using `zquantum.qaoa.ansatzes.farhi_ansatz`, so renaming it will probably break something somewhere. I guess I'll rename the new ansatz file but not the old ones?

**Review by mstechly — APPROVED** [review]
